src/libcprover-cpp/add_dependencies.sh

```diff
@@ -15,11 +15,36 @@ DESTINATION=$1
 shift
 LIB_LIST=$@
 
+# We need to do the following renaming of object files because of translation
+# unit name collisions which affect linking against the final artefact. For more
+# details, please look at https://github.com/diffblue/cbmc/issues/7586 .
+
 # For each library to add:
 for lib in ${LIB_LIST}; do
+  # We will unpack and rename all .o of dependent libraries marking them with
+  # their library name to avoid clashes.
+  LIBNAME=$(basename ${lib} .a)
+
+  # Remove previous unpacked folders and create a new fresh one to work in.
+  rm -rf ${LIBNAME}
+  mkdir ${LIBNAME}
+  cd ${LIBNAME}  
+
   # Unpack the library
   ${AR_COMMAND} -x ${lib}
+
+  # Rename all object file in the library prepending "${LIBNAME}_" to avoid
+  # clashes.
+  for obj in *.o; do
+    mv ${obj} "${LIBNAME}_${obj}"
+  done
+
+  # Move back to the working directory.
+  cd ..
 done
 
 # Append all the unpacked files to the destination library
-${AR_COMMAND} -rcs ${DESTINATION} *.o
+${AR_COMMAND} -rcs ${DESTINATION} **/*.o
```

Comment:
Why `**`?

Reply:
`**` supports matching the pattern across multiple directories, which is needed because of the preceding changes which places the `.o` files in separate directories.

Reply:
Because the files are in the subfolder we need to add all `.o` in the subfolders.
However this can be problematic, so Fotis is now pushing an updated and more robust version of this

Reply:
It would have worked if you used 1) `bash` as the shell and 2) had the `globstar` option set (for those versions of `bash` that even support it). As it was, `**` was not meaningful. But then you've now found a different solution.